전남대 BE_김보민_2주차 과제 (1단계)

KBM05haruin30 opened the pull request:
### 코드 작성하면서 어려웠던 점
validation을 처음 사용해 봐서 참고 자료를 보면서 했습니다. 참고 자료로 못 찾은 내용들은 구글링을 통해 찾아서 사용하였습니다. 처음 사용하는 것이다 보니 헷갈리는 것들이 있었습니다. 또한, 중복된 로직을 제거하면서 제가 생각한 순서랑 다르게 흘러가는 느낌이 있었습니다.

### 궁금한 점
1. 중복된 로직을 사용하기 위해 새로 메서드를 만들어 코드를 작성했습니다. 하지만 제가 생각한 순서랑 다르게 흘러가는 것 같습니다. 상품을 추가할 때 가격에 숫자가 아닌 값만 넣으면 다른 곳에서는 예외 처리 메시지가 발생하지 않습니다.
![캡처 PNG1231](https://github.com/kakao-tech-campus-2nd-step2/spring-gift-wishlist/assets/138747411/2c51a81d-573e-4e90-b65c-c1d35cded7d4)

2. 제가 작성한 코드는 클라이언트가 잘못된 값을 입력하면 입력하는 칸 밑에 예외 처리 메시지가 나옵니다. GlobalExceptionHandler에서 @ResponseStatus를 사용하여 예외 처리를 하면 400 Bad Request를 반환하도록 했습니다. 하지만 크롬에서 개발자 도구를 킨 다음 잘못된 값을 입력해보면 http status 400이 보이지 않고 200이 보입니다. 이렇게 보이는게 정상인지 궁금합니다. 만약 정상이라면 400은 어디서 확인할 수 있는지도 궁금합니다.

KBM05haruin30 committed:
docs : 1단계 구현할 기능 목록 정리
1단계 - 유효성 검사 및 예외 처리를 하기 위해 필요한 구현 기능 목록을 정리했습니다.

KBM05haruin30 committed:
feat : 상품 이름 길이 검사
상품 이름은 최소 1자 이상이어야 하므로 @NotBlank를 사용했습니다. 그리고 공백 포함 최대 15자까지 입력 가능하므로 @Size(max=15)를 사용했습니다.

KBM05haruin30 committed:
feat : 상품 이름 특수 문자 검사
@Pattern을 사용했습니다. 상품 이름에 (), [], +, -, &, /, _ 외 특수 문자는 사용 불가능하므로 사용 가능한 문자는 regexp에 넣습니다.

KBM05haruin30 committed:
feat : 상품 가격 검사
@Pattern을 사용했습니다. 상품 이름에 (), [], +, -, &, /, _ 외 특수 문자는 사용 불가능하므로 사용 가능한 문자는 regexp에 넣습니다.

KBM05haruin30 committed:
feat : 상품 이미지 주소 검사
없으면 안되므로 @NotBlank를 사용했습니다.
주소의 특징은 https://~~~ 또는 http://~~~이므로 @Pattern 사용했습니다.

KBM05haruin30 committed:
feat : Product 앞에 @Valid 사용
@Valid 어노테이션을 사용하면, Spring은 자동으로 요청 객체의 필드에 대한 유효성을 검사합니다. Product 필드에 유효성 검증 어노테이션이 적용되어 있으므로 @Valid를 사용합니다.

KBM05haruin30 committed:
refactor : 사용하지 않는 메서드 삭제

KBM05haruin30 committed:
feat : 카카오 문구 검사
입력 받은 이름에서 "카카오"가 있으면 예외 발생합니다.

KBM05haruin30 committed:
feat : 전체적인 예외 다루는 새로운 클래스 생성
전체적인 예외를 다루기 위해 새로운 클래스인 GlobalExceptionHandler를 만들었습니다.

KBM05haruin30 committed:
docs : 1단계 구현할 기능 목록 수정

KBM05haruin30 committed:
feat : MethodArgumentNotValidException 예외 처리

KBM05haruin30 committed:
feat : IllegalArgumentException 예외 처리

KBM05haruin30 committed:
feat : @ResponseStatus 사용
@ResponseStatus 사용하여 Http Status Code를 반환합니다.

KBM05haruin30 committed:
refactor : 코드 정리

KBM05haruin30 committed:
docs : 1단계 구현할 기능 목록 수정

KBM05haruin30 committed:
feat : 카카오 문구 검사 사용

KBM05haruin30 committed:
docs : 1단계 구현할 기능 목록 수정

KBM05haruin30 committed:
docs : 1단계 구현할 기능 목록 수정

KBM05haruin30 committed:
feat : BindingResult 추가
BindingResult는 폼 데이터가 객체에 바인딩되는 과정에서 발생하는 오류를 처리합니다. 그리고 @Valid 어노테이션을 사용하여 수행된 유효성 검사에서 발생하는 오류를 처리합니다.

KBM05haruin30 committed:
feat : 상품 가격에 숫자 외의 값 입력 시 예외 처리
상품 가격은 숫자만 올 수 있습니다. 숫자 외의 값이 올 경우 BindingResult를 사용하여 예외를 처리합니다.

KBM05haruin30 committed:
feat : 카카오 문구 검사 예외 처리
카카오 문구 검사를 통해 예외를 던진 것을 try-catch로 받아서 처리합니다.

KBM05haruin30 committed:
refactor : 중복된 로직 제거
중복된 로직을 하나의 메서드로 만들어서 사용했습니다.

KBM05haruin30 committed:
refactor : @ControllerAdvice로 변경

KBM05haruin30 committed:
refactor : 카카오 문구 검사의 모델 속성 값 수정
카카오 문구 검사 메서드의 Model의 attributeName을 더욱 직관성있게 수정했습니다. 수정하면서 이 예외를 처리하는 GlobalExceptionHandler의 메서드 부분도 수정해주었습니다.

KBM05haruin30 committed:
feat : add_product_form 수정
예외 처리 결과 메시지를 표시하도록 html 파일을 수정합니다.

KBM05haruin30 committed:
feat : add_product_form 수정
button의 type과 표시될 이름이 반대로 적혀있었습니다.

KBM05haruin30 committed:
feat : edit_product_form 수정
button의 type과 표시될 이름이 반대로 적혀있었습니다.

KBM05haruin30 committed:
feat : edit_product_form 수정
예외 처리 결과 메시지를 표시하도록 html 파일을 수정합니다.

KBM05haruin30 committed:
feat : addProduct 수정

KBM05haruin30 committed:
refactor : 코드 정리

KBM05haruin30 committed:
feat : ProductDTO 생성
validation 코드와 모델링 코드를 분리하기 위해서 DTO 클래스를 만들었습니다.

KBM05haruin30 committed:
feat : ProductConverter 생성
Product 엔티티와 ProductDTO 간의 변환을 처리하기 위해 ProductConverter를 만들었습니다.

KBM05haruin30 committed:
refactor : Product 수정
ProductDTO를 만들었으므로 validation 코드는 DTO에 포함이 됩니다. 그러므로 엔티티에서는 validation 코드를 지워줍니다.

KBM05haruin30 committed:
refactor : AdminController 수정
ProductDTO를 사용하도록 수정해줍니다.

KBM05haruin30 committed:
refactor : html 수정
AdminController에서 Model의 attributeName을 "product"에서 "productDTO"로 변경했으므로 html에서도 productDTO를 쓰도록 변경한다.

KBM05haruin30 committed:
feat : @Pattern 정규식 수정
정규식에서 자음과 모음이 따로 나와있는 것도 인식되게 수정하였습니다.

KBM05haruin30 committed:
refactor : 코드 정리

KBM05haruin30 committed:
feat : @Pattern 추가
카카오 단어가 포함된 문구를 처리하기 위해서 @Pattern을 사용했습니다.

KBM05haruin30 committed:
refactor : @Pattern 추가로 인해 수정
@Pattern 추가로 인해 필요 없는 부분을 수정했습니다.

Hongdonggeon commented on src/main/java/gift/Model/ProductDTO.java:
```diff
@@ -0,0 +1,20 @@
+import jakarta.validation.constraints.PositiveOrZero;
+import jakarta.validation.constraints.Size;
+
+public record ProductDTO(Long id,
```
화면에서 id 값을 받지 않고 있어서, Dto에 id 필드가 있을 필요가 없을 것 같아요!

KBM05haruin30 replied:
넵! Product에서 ProductDTO로 복사하면서 지우는 것을 깜빡 잊어버린 것 같습니다. 바로 수정하였습니다!

Hongdonggeon commented on src/main/java/gift/Controller/ProductConverter.java:
```diff
@@ -0,0 +1,15 @@
+public class ProductConverter {
+    public static ProductDTO toDTO(Product product) {
+        return new ProductDTO(product.id(), product.name(), product.price(), product.imageUrl());
+    }
```
ProductConverter 객체도 좋지만, 요구사항이 많아지고 다루는 객체가 많아진다면 모든 객체에 대해 Converter를 만들긴 힘들거예요 
이 메서드는 ProductDto 클래스안에 만들어도 되지 않을까요?

KBM05haruin30 replied:
record 클래스 안에 메서드를 작성할 수 없는 줄 알았었습니다. ProductDTO 클래스 안에 만들어보겠습니다!

KBM05haruin30 replied:
수정 완료했습니다!

Hongdonggeon commented on src/main/java/gift/Model/ProductDTO.java:
```diff
@@ -0,0 +1,20 @@
+import jakarta.validation.constraints.PositiveOrZero;
+import jakarta.validation.constraints.Size;
+
+public record ProductDTO(Long id,
```
Dto 클래스가 모델 패키지 안에 있네요? Dto는 컨트롤러에서만 사용하는데 모델 패키지 보다 컨트롤러 패키지가 더 잘 어울리지 않을까요 ? 
Dto의 역할과 개념에 대해서 공부해보면 좋을 것 같고, 이번 기회에 Dto vs 도메인(모델) 에 대해 공부해봐도 좋을 것 같아요
참고할만한 자료도 공유드려요
https://tecoble.techcourse.co.kr/post/2021-04-25-dto-layer-scope/
https://tecoble.techcourse.co.kr/post/2021-05-16-dto-vs-vo-vs-entity/

KBM05haruin30 replied:
자료 공유 감사합니다! 읽어보니 도움이 된 것 같습니다. 그리고 DTO-Entity간의 변환 위치는 Service 레이어에서 하는 것이 타당한 것 같습니다. 바로 위의 코드 수정은 ProductDTO에 DTO와 Entity간의 변환 로직이 추가되어 있으므로 이를 제거하고 전송에만 더 집중할 수 있도록 만들고, Service 계층을 만들어야 할 것 같습니다. 추가로 DTO를 컨트롤러 패키지가 아닌 dto 패키지를 새로 만들어서 거기에 넣어도 되는지 궁금합니다.

Hongdonggeon replied:
여러 패키지에서 공유해야 하는 상황이면 보민님이 말하신 것 처럼 별도의 dto 패키지를 만들어서 넣어두는 것도 좋다고 생각하고요!
만약 서비스 계층까지만 사용되면, 해당 서비스 계층에 dto 패키지를 만드는 방법도 있다고 들었습니다!
고민해보시고 결정하시면 될 것 같습니다!

Hongdonggeon commented on src/main/java/gift/Controller/AdminController.java:
```diff
@@ -30,27 +33,37 @@ public String listProducts(Model model) {
     @PostMapping("/add")
-    public String addProduct(@ModelAttribute Product product, Model model) {
-        productRepository.saveProduct(product);
+    public String addProduct(@ModelAttribute @Valid ProductDTO productDTO, BindingResult result,
```
BindingResult 를 쓰셨네요!  BindingResult를 선언하지 않고 `@Valid`에서 발생한 예외를 GlobalExceptionHandler에서 처리하는 방법도 있었을텐데 BindingResult를 사용한 방법이 궁금합니다!

KBM05haruin30 replied:
1단계 과제에서 힌트로 주신 스프링 프레임워크 문서를 봤더니 Controller부분에서 BindingResult를 사용하여 사용해야 되는 줄 알고 사용했습니다. BindingResult를 선언하지 않고 예외를 처리하는 방법을 찾아서 수정해보도록 하겠습니다!

KBM05haruin30 replied:
그리고 저는 오류가 발생하면 현재 머무르고 있는 페이지에 값이 입력된 채로 가게 했습니다. AdminController에서 BindingResult를 선언하지 않고 예외를 GlobalExceptionHandler에서 처리하는 방법이 있다고 말씀해주셨는데 그럼 GlobalExceptionHandler에서도 BindingReuslt를 선언하지 않아도 되는 것인지 궁금합니다.

KBM05haruin30 replied:
구글링을 통해 찾아보고 있는데 감이 안잡힙니다..

Hongdonggeon replied:
GlobalExcetpionHandler에서는 BindingResult를 선언해야합니다! 
`@Valid`에서 예외가 발생하면 MethodArgumentNotValidException 클래스로 발생하게 될거예요
GlobalExceptionHandler에서 저 예외 클래스에 대해서 예외를 공통으로 처리하는 방법을 추천드린겁니다!!

```    
    @ExceptionHandler({MethodArgumentNotValidException.class})
    public ResponseEntity<ErrorResponse> handleInvalidRequest(final BindingResult bindingResult) {
        final List<FieldError> fieldErrors = bindingResult.getFieldErrors();
           ~~~~~
        return ResponseEntity.badRequest().body();
    }
```

Hongdonggeon commented on src/main/java/gift/Controller/AdminController.java:
```diff
@@ -60,7 +73,14 @@ public String deleteProduct(@PathVariable("id") long id) {
+    private boolean validateProduct(ProductDTO productDTO, BindingResult result, Model model) {
+        if (result.hasErrors()) {
+            if (result.hasFieldErrors("price")) {
+                model.addAttribute("priceError", "가격은 숫자만 입력 가능합니다.");
```
가격에 대한 예외 메시지만 이곳에서 처리한 이유가 있을까요?

KBM05haruin30 replied:
validation annotation으로 처리하고 싶었습니다. 그래서 @Pattern을 이용해서 처리를 해볼까 생각했지만 @Patttern은 String만 가능하다고 하여 따로 예외 메시지를 처리했습니다. 저는 ProductDTO에서 long price로 사용하였는데 이것을 String price로 바꾸고 따로 long price로 변경을 해도 문제가 없는지 궁금합니다.

KBM05haruin30 replied:
위에서 record에 메서드를 추가할 수 있다는 것을 깨달았습니다. ProductDTO의 price를 String으로 변경한 후 @Pattern으로 숫자만 입력 받게 지정해주면 가격에 대한 예외 메시지만 따로 처리하지 않아도 될 것 같다는 생각이 들어서 그렇게 수정해보려고 합니다!

KBM05haruin30 replied:
수정 완료 했습니다!

Hongdonggeon commented on src/main/java/gift/Model/Product.java:
```diff
@@ -1,5 +1,8 @@
+public record Product(Long id,
+                      String name,
+                      long price,
+                      String imageUrl) {
```
모델에서 검증은 안해도될까요?

KBM05haruin30 replied:
모델은 비즈니스 로직만 집중하고 DTO에서 검증을 하는 것이라 생각하고 나눴었습니다. 하지만 검색해서 찾아보니 둘 다 사용하는 것이 좋다고 나와있는 것을 확인했습니다. 모델에서도 검증을 하도록 수정하겠습니다! 하지만 궁금한 점이 있습니다. DTO랑 엔티티 둘 다 검증을 넣게 되면 코드가 복잡해지는건 아닌지 궁금합니다.

Hongdonggeon replied:
모델에 검증로직이 있다고 코드가 복잡해진다고 생각하지 않습니다!

Hongdonggeon reviewed (commented):
.

Hongdonggeon reviewed (changes requested):
보민님 안녕하세요  
미션 진행하느라 수고하셨습니다! 
기본 요구사항에 대해서는 잘 구현하셨어요 👍
지금은 요구사항이 작고 간단해서 괜찮지만, 서비스 계층을 추가해서 검증 로직을 작성하는 것도 좋을 것 같아요! 
그리고 간단한 코멘트들  남겨놓았습니다! 반영하면서 궁금한 것 있으시면 질문주세요!

KBM05haruin30 committed:
refactor : ProductDTO 수정
화면에서 입력 받을 때 id를 사용하지 않으므로 id를 삭제했습니다.

KBM05haruin30 committed:
feat : Product 수정
Product에서도 검증을 할 수 있게 수정하였습니다.

KBM05haruin30 committed:
feat : ProductConverter 삭제 및 ProductDTO 수정
요구 사항이 많아지고 다루는 객체가 많아지면 모든 객체에 대해 Converter를 만들기 힘들어집니다. 그래서 ProductConverter를 삭제하고 그 안에 있던 내용을 ProductDTO에 넣었습니다.

KBM05haruin30 committed:
feat : price 검증을 위해 타입 변경
price 검증을 위해 long에서 String으로 변경하여 @Pattern을 사용해 숫자만 입력 가능하도록 변경했습니다. 그리고 그에 맞게 메서드를 작성 및 수정했습니다.

KBM05haruin30 committed:
feat : price 타입 변경으로 인한 수정

KBM05haruin30 committed:
feat : AdminController에서 price 예외 처리 제거
이제 price는 validation annotation으로 검증을 하므로 필요 없어진 메서드를 삭제했습니다. 그에 따라 html 파일의 불필요한 부분도 삭제했습니다.

KBM05haruin30 committed:
refactor : ProductDTO 컨트롤러 패키지로 이동 및 ProductDTO의 로직 제거
ProductDTO는 DTO이므로 로직을 제거했습니다.

KBM05haruin30 committed:
feat : Service 계층 만들기
ProductDTO에 있던 로직을 Service에 넣었습니다.

KBM05haruin30 committed:
feat : 수정하면서 삭제된 것 복구
수정하면서 @PostMapping에서 model.addAttribute("productDTO", productDTO);를 잘못 지워서 다시 썼습니다.

KBM05haruin30 committed:
refactor : 사용하지 않는 메서드 삭제
사용할 생각으로 만들었는데 사용할 필요가 없어져 삭제했습니다.

KBM05haruin30 committed:
refactor : final 추가
ProductRepository 앞에 final을 추가했습니다.

Hongdonggeon reviewed (approved):
안녕하세요 보민님! 
미션 진행하느라 리뷰 반영하느라 수고많으셨습니다! 
열심히 학습하신 모습도 보기 좋습니다 
다음 미션도 화이팅입니다!
